File: NOVEMBER-2025/22-11-2025/Minimum Operations to Connect Hospitals.cpp
class Solution {
public:
    vector<int> parent, rankv;
    int find(int x) {
        if (parent[x] != x)
            parent[x] = find(parent[x]);
        return parent[x];
    }
    void unite(int x, int y) {
        int px = find(x), py = find(y);
        if (px == py) return;
        
        if (rankv[px] < rankv[py]) {
            parent[px] = py;
        } else if (rankv[px] > rankv[py]) {
            parent[py] = px;
        } else {
            parent[py] = px;
            rankv[px]++;
        }
    }
    int minConnect(int V, vector<vector<int>>& edges) {
        parent.resize(V);
        rankv.resize(V, 0);
        for (int i = 0; i < V; i++)
            parent[i] = i;
        int redundant = 0;
        for (auto &e : edges) {
            int u = e[0], v = e[1];
            if (find(u) == find(v)) {
                redundant++;
            } else {
                unite(u, v);
            }
        }
        int components = 0;
        for (int i = 0; i < V; i++) {
            if (find(i) == i)
                components++;
        }
        int needed = components - 1;
        if (redundant >= needed) return needed;
        return -1;
    }
};

File: NOVEMBER-2025/22-11-2025/Minimum Operations to Connect Hospitals.js
/**
 * @param {number} V
 * @param {number[][]} edges
 * @returns {number}
 */

class Solution {
    minConnect(V, edges) {

        const parent = Array.from({ length: V }, (_, i) => i);
        const rank = Array(V).fill(0);

        const find = (x) => {
            if (parent[x] !== x) parent[x] = find(parent[x]);
            return parent[x];
        };

        const unite = (a, b) => {
            let pa = find(a), pb = find(b);
            if (pa === pb) return;

            if (rank[pa] < rank[pb]) parent[pa] = pb;
            else if (rank[pb] < rank[pa]) parent[pb] = pa;
            else {
                parent[pb] = pa;
                rank[pa]++;
            }
        };

        let redundant = 0;

        for (let [u, v] of edges) {
            if (find(u) === find(v)) redundant++;
            else unite(u, v);
        }

        let components = 0;
        for (let i = 0; i < V; i++) {
            if (find(i) === i) components++;
        }

        let needed = components - 1;
        return redundant >= needed ? needed : -1;
    }
}